interfaces/builtin/rtc.go

```diff
@@ -0,0 +1,115 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/snapcore/snapd/interfaces"
+)
+
+const rtcControlConnectedPlugAppArmor = `
+# Description: Allow write access to system real-time clock
+# See 'man 4 rtc' for details.
+
+/dev/rtc rw,
+/dev/rtc[0-9]* rw,
+
+# As the core snap ships the hwclock utility we can also allow
+# clients to use it now that they have access to the relevant
+# device nodes.
+/sbin/hwclock ixr,
+`
+
+// The type for the rtc interface
+type RtcInterface struct{}
+
+// Getter for the name of the rtc interface
+func (iface *RtcInterface) Name() string {
+	return "rtc"
+}
+
+func (iface *RtcInterface) String() string {
+	return iface.Name()
+}
+
+// Check validity of the defined slot
+func (iface *RtcInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	// Does it have right type?
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	// Creation of the slot of this type
+	// is allowed only by a gadget or os snap
+	if !(slot.Snap.Type == "os") {
+		return fmt.Errorf("%s slots only allowed on core snap", iface.Name())
+	}
+	return nil
+}
+
+// Checks and possibly modifies a plug
+func (iface *RtcInterface) SanitizePlug(plug *interfaces.Plug) error {
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface))
+	}
+	// Currently nothing is checked on the plug side
+	return nil
+}
+
+// Returns snippet granted on install
+func (iface *RtcInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
+
+// Getter for the security snippet specific to the plug
+func (iface *RtcInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		return []byte(rtcControlConnectedPlugAppArmor), nil
+
+	case interfaces.SecurityUDev:
+		var tagSnippet bytes.Buffer
+		const udevRule = `KERNEL=="/dev/rtc", TAG+="%s"`
```

Comment:
We have to use udev tagging as some of our interfaces do and we can't leave others not using them as then the user would end up with an AppArmor rule allowing access to /dev/rtc but it's not in /dev because he also uses another interface which uses udev tagging. Some of the old interfaces (like tpm) don't do this yet but that is something we have to change too.